[c10d]Prototype of remote_group_merge

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #158287
* #157716

Tentative implementation of merge_remote_group per the proposal here: [docs.google.com/document/d/13R-1t_yESTvmAjcCN-wQjQQadIEu0JNIdS65uZawZzY/edit?tab=t.0#heading=h.3ctbqqopzc89](https://docs.google.com/document/d/13R-1t_yESTvmAjcCN-wQjQQadIEu0JNIdS65uZawZzY/edit?tab=t.0#heading=h.3ctbqqopzc89)


cc @H-Huang @awgu @wanchaol @fegin @wz337 @wconstab @d4l3k

fduwjj committed:
[c10d]Prototype of remote_group_merge

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d]Prototype of remote_group_merge"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d]Prototype of remote_group_merge"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d]Prototype of remote_group_merge"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d]Prototype of remote_group_merge"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroup.cpp:
```diff
@@ -216,6 +215,58 @@ c10::intrusive_ptr<ProcessGroup> ProcessGroup::splitGroup(
 }
 
+std::vector<uint8_t> toByteVector(uint64_t value) {
+  std::vector<uint8_t> bytes(8);
+  for (int i = 0; i < 8; ++i) {
+    bytes[i] = static_cast<uint8_t>((value >> (i * 8)) & 0xFF);
+  }
```
Opps actually Let me remove these two utils functions which I don't end up using.

fduwjj committed:
Update on "[c10d]Prototype of remote_group_merge"


Tentative implementation of merge_remote_group per the proposal here: [docs.google.com/document/d/13R-1t_yESTvmAjcCN-wQjQQadIEu0JNIdS65uZawZzY/edit?tab=t.0#heading=h.3ctbqqopzc89](https://docs.google.com/document/d/13R-1t_yESTvmAjcCN-wQjQQadIEu0JNIdS65uZawZzY/edit?tab=t.0#heading=h.3ctbqqopzc89)


cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d]Prototype of remote_group_merge"


Tentative implementation of merge_remote_group per the proposal here: [docs.google.com/document/d/13R-1t_yESTvmAjcCN-wQjQQadIEu0JNIdS65uZawZzY/edit?tab=t.0#heading=h.3ctbqqopzc89](https://docs.google.com/document/d/13R-1t_yESTvmAjcCN-wQjQQadIEu0JNIdS65uZawZzY/edit?tab=t.0#heading=h.3ctbqqopzc89)


cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d]Prototype of remote_group_merge"


Tentative implementation of merge_remote_group per the proposal here: [docs.google.com/document/d/13R-1t_yESTvmAjcCN-wQjQQadIEu0JNIdS65uZawZzY/edit?tab=t.0#heading=h.3ctbqqopzc89](https://docs.google.com/document/d/13R-1t_yESTvmAjcCN-wQjQQadIEu0JNIdS65uZawZzY/edit?tab=t.0#heading=h.3ctbqqopzc89)


cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

d4l3k commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -3066,6 +3076,28 @@ options :class:`~torch.distributed.ProcessGroupNCCL.Options`).
           .def_readwrite("_timeout", &::c10d::Backend::Options::timeout);
 
+  intrusive_ptr_class_<::c10d::ProcessGroup::MergeOptions>(
+      module, "MergeOptions", R"(Class for ProcessGroup Merge Options)")
```
Thoughts on taking these directly as arguments to merge_remote_group instead of having a separate object? 

I think it's generally nicer for the Python API to avoid the options object though as a hybrid we can expose kwargs at the Python layer and options in C++? Same way we handle the other PG op options

d4l3k reviewed (approved):
This is looking good! There's a few things I think we can clean up before landing

fduwjj commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -2063,6 +2064,7 @@ communication mechanism.
           .def("rank", &::c10d::ProcessGroup::getRank, R"(Get the rank of this process group.)")
           .def("size", &::c10d::ProcessGroup::getSize, R"(Get the size of this process group.)")
           .def("name", &::c10d::ProcessGroup::getBackendName, R"(Get the name of this process group.)")
+          .def("store", &::c10d::ProcessGroup::getStore, R"(Get the store of this process group.)")
```
OK, I will remove this one.

fduwjj commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -2071,6 +2073,14 @@ communication mechanism.
+              &::c10d::ProcessGroup::mergeRemoteGroup,
+              py::arg("store"),
+              py::arg("merge_options"),
+              py::arg("rank"),
```
I have thought about this for a long time since yesterday but have not figured out a good decent way.

fduwjj committed:
Update on "[c10d]Prototype of remote_group_merge"


Tentative implementation of merge_remote_group per the proposal here: [docs.google.com/document/d/13R-1t_yESTvmAjcCN-wQjQQadIEu0JNIdS65uZawZzY/edit?tab=t.0#heading=h.3ctbqqopzc89](https://docs.google.com/document/d/13R-1t_yESTvmAjcCN-wQjQQadIEu0JNIdS65uZawZzY/edit?tab=t.0#heading=h.3ctbqqopzc89)


cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj commented:
Addressed all comments from reviewers.

fduwjj committed:
Update on "[c10d]Prototype of remote_group_merge"


Tentative implementation of merge_remote_group per the proposal here: [docs.google.com/document/d/13R-1t_yESTvmAjcCN-wQjQQadIEu0JNIdS65uZawZzY/edit?tab=t.0#heading=h.3ctbqqopzc89](https://docs.google.com/document/d/13R-1t_yESTvmAjcCN-wQjQQadIEu0JNIdS65uZawZzY/edit?tab=t.0#heading=h.3ctbqqopzc89)


cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge

fduwjj commented:
@pytorchbot merge -i
